warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java

```diff
@@ -112,6 +114,21 @@ private JexlNode flattenTree(Deque<JexlNode> postOrderStack) {
             // if this is an AND or OR node, flatten it
             else if (hasChildren && (node instanceof ASTOrNode || node instanceof ASTAndNode)) {
                 newNode = flattenAndOrNode(JexlNodes.children(parentStack.pop(), childrenStack.pop().toArray(new JexlNode[0])));
+                
+                // if the returned node is not a junction node, then we pruned. check for an extra reference expression.
+                //  @formatter:off
+                if (!(newNode instanceof ASTAndNode || newNode instanceof ASTOrNode) &&
+                        postOrderStack.size() >= 3 &&
+                        postOrderStack.peek() instanceof ASTReferenceExpression) {
+                    //  formatter:on
+                    JexlNode t1 = postOrderStack.pop();
+                    if (postOrderStack.peek() instanceof ASTReference) {
+                        postOrderStack.pop();
+                        pruned = true;
+                    } else {
+                        postOrderStack.push(t1);
+                    }
+                }
```

Comment:
The ability to remove superfluous references and reference expressions already exists, and is toggleable via the `removeReferences` boolean.  See `TreeFlatteningRebuildingVisitor.flattenAll`.

Reply:
The `flattenAll` method has a warning that it may adversely affect the evaluation of a query. 

I redirected the `flatten` method to `flattenAll` and got a bunch of range and geo query failures. After some digging it looks like the `flattenAll` will unwrap the source of a marked bounded range as well as unmarked bounded ranges.

Adding a few lines to the `TreeFlatteningRebuildingVisitorTest` provided the following examples
```
# marked bounded range
((_Bounded_ = true) && (STATE >= 'e' && STATE <= 'r'))
((_Bounded_ = true) && STATE >= 'e' && STATE <= 'r')

# unmarked bounded range (maybe this is fine)
GEO == '1f36c71c71c71c71c7' && (WKT_BYTE_LENGTH >= '+AE0' && WKT_BYTE_LENGTH < '+bE8')
GEO == '1f36c71c71c71c71c7' && WKT_BYTE_LENGTH >= '+AE0' && WKT_BYTE_LENGTH < '+bE8'
```

There's also a few problems with the input queries of the TreeFlatteningRebuildingVisitorTest where the input query structure will change during the initial parse. I think we should have the inputs match the result of `JexlStringBuildingVisitor.buildQueryWithoutParse(JexlASTHelper.parseQuery("query string"))`, let me know what you think.

All that to say, I don't think `flattenAll` should be used to address the specific case of leftover parens after a junction node has been pruned. At least not in it's current state.

Reply:
I'm not advocating for `flattenAll` being used in place of `flatten` in all cases.  In this case though (query pruning), that seems like the functionality you need.  I still question why we aren't updating the QueryPruningVisitor to have it properly prune these nodes instead.  

The comment stating that 'removing reference expressions and references may adversely affect the evaluation of a query' was carried over from the original implementation: (https://github.com/NationalSecurityAgency/datawave/blob/1.1.0/warehouse/query-core/src/main/java/datawave/query/rewrite/jexl/visitors/TreeFlatteningRebuildingVisitor.java#L42)  Maybe that was true at the time, but I don't think that's true today.

My understanding of how the tree flattening rebuilding visitor is supposed to work is that `flatten` leaves parens as-is, while `flattenAll` collapses nested parens, and removes unnecessary parens where applicable, but 1) doesn't affect the evaluation of the tree, and 2) doesn't touch any marked or negated nodes (see `TreeFlatteningRebuilder.flattenReference` to see the evidence of that).  So, by adding logic that conditionally removes parens in the `flatten` call, we would be breaking that contract.

I honestly think we already have the functionality you need in the `flattenAll` call, but I am still advocating for fixing QueryPruningVisitor instead.

Reply:
I'm good with backing out the changes in the TreeFlatteningRebuilder and fixing the QueryPruningVisitor to clean up junctions with single nodes leftover from a prune.

Per our discussion I'll look at cleaning up the TreeFlatteningRebuilderTest to exercise both flatten and flattenAll cases.